framework/src/config/config_reader.rs

```diff
@@ -13,297 +14,109 @@ pub const DEFAULT_NAME: &'static str = "zcsi";
-                    "Could not parse loopback spec {:?}",
-                    v
-                )).into())
-            }
-        };
-
-        let tso = match port_def.get("tso") {
-            Some(&Value::Boolean(l)) => l,
-            None => false,
-            v => {
-                return Err(ErrorKind::ConfigurationError(format!(
-                    "Could not parse tso spec {:?}",
-                    v
-                )).into())
-            }
-        };
-
-        let csum = match port_def.get("checksum") {
-            Some(&Value::Boolean(l)) => l,
-            None => false,
-            v => {
-                return Err(ErrorKind::ConfigurationError(format!(
-                    "Could not parse csum spec {:?}",
-                    v
-                )).into())
-            }
-        };
-
-        let symmetric_queue = port_def.contains_key("cores");
-        if symmetric_queue
-            && (port_def.contains_key("rx_cores") || port_def.contains_key("tx_cores"))
-        {
-            println!(
-                "cores specified along with rx_cores and/or tx_cores for port {}",
-                name
-            );
-            return Err(ErrorKind::ConfigurationError(format!(
-                "cores specified along with rx_cores and/or tx_cores \
-                 for port {}",
-                name
-            )).into());
-        }
+static DEFAULT_TOML: &'static str = r#"
+    duration = 0
+    remotes = []
+    agents = []
+
+    [dpdk]
+      name = "zcsi"
+      secondary = false
+      primary_core = 0
+      cores = [0]
+      strict = false
+      pool_size = 2047
+      cache_size = 32
+      ports = []
+"#;
+
+#[derive(Debug, Deserialize)]
+pub struct OccamConfiguration {
+    pub duration: u64,
+    pub remotes: Vec<RemoteConfiguration>,
+    pub agents: Vec<AgentConfiguration>,
+    pub dpdk: NetbricksConfiguration,
+}
 
-        fn read_queue(queue: &Value) -> Result<Vec<i32>> {
-            match *queue {
-                Value::Array(ref queues) => {
-                    let mut qs = Vec::with_capacity(queues.len());
-                    for q in queues {
-                        if let Value::Integer(core) = *q {
-                            qs.push(core as i32)
-                        } else {
-                            return Err(ErrorKind::ConfigurationError(format!(
-                                "Could not parse queue spec {:?}",
-                                q
-                            )).into());
-                        };
-                    }
-                    Ok(qs)
-                }
-                Value::Integer(core) => Ok(vec![core as i32]),
-                _ => Ok(vec![]),
-            }
+impl Default for OccamConfiguration {
+    fn default() -> OccamConfiguration {
+        OccamConfiguration {
+            duration: 0,
+            remotes: vec![],
+            agents: vec![],
+            dpdk: Default::default(),
```

Comment:
What does implementing Default for this buy us? Honestly kinda seems like we don't want inline defaults, but want to require that everything is specified in the default config and then overridden by the real config. Having these kinds of inline defaults that in practice can never see the light of day is mostly confusing IMO.

Same question goes for the remote / agent definitions, though those seem a little different as they contain mutable state (healthy or no). That state will move out of there pretty quickly, though, and for the PoC we'll certainly be specifying the health of each agent involved explicitly.

Reply:
Ok, yeah that's the behavior here as well. But then how do you handle defaults in the array subtype. If I want these things to be equal, can this be done hierarchically? I think the answer is no here.

```
[ { bar = 300 } ] = [ { bar = 300, baz = fale } ]
```